src/api/procedures/addInstruction.ts

```diff
@@ -228,40 +318,43 @@ export async function getAuthorization(
  * @hidden
  */
 export async function prepareStorage(
-  this: Procedure<Params, Instruction, Storage>,
-  { legs }: Params
+  this: Procedure<Params, Instruction[], Storage>,
+  { instructions }: Params
 ): Promise<Storage> {
   const { context } = this;
 
   const identity = await context.getCurrentIdentity();
 
-  const portfolios = await P.map(legs, async ({ from, to }) => {
-    const fromPortfolio = portfolioLikeToPortfolio(from, context);
-    const toPortfolio = portfolioLikeToPortfolio(to, context);
+  const portfoliosToAffirm = await P.map(instructions, async ({ legs }) => {
+    const portfolios = await P.map(legs, async ({ from, to }) => {
+      const fromPortfolio = portfolioLikeToPortfolio(from, context);
+      const toPortfolio = portfolioLikeToPortfolio(to, context);
 
-    const result = [];
-    const [fromCustodied, toCustodied] = await Promise.all([
-      fromPortfolio.isCustodiedBy({ identity }),
-      toPortfolio.isCustodiedBy({ identity }),
-    ]);
+      const result = [];
+      const [fromCustodied, toCustodied] = await Promise.all([
+        fromPortfolio.isCustodiedBy({ identity }),
+        toPortfolio.isCustodiedBy({ identity }),
+      ]);
 
-    if (fromCustodied) {
-      result.push(fromPortfolio);
-    }
+      if (fromCustodied) {
+        result.push(fromPortfolio);
+      }
 
-    if (toCustodied) {
-      result.push(toPortfolio);
-    }
+      if (toCustodied) {
+        result.push(toPortfolio);
+      }
 
-    if (result.length) {
-      return result;
-    }
+      if (result.length) {
+        return result;
+      }
 
-    return undefined;
+      return [];
```

Comment:
Now that you always return an array you can just replace all this with `return result;`